File: notification-service/src/main/java/com/fiapql/notification/config/RabbitMqConfig.java
package com.fiapql.notification.config;

import org.springframework.amqp.core.*;
import org.springframework.amqp.rabbit.config.SimpleRabbitListenerContainerFactory;
import org.springframework.amqp.rabbit.connection.ConnectionFactory;
import org.springframework.amqp.support.converter.Jackson2JsonMessageConverter;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;

@Configuration
public class RabbitMqConfig {

    /** DLQ declarada aqui também para garantir que exista ao subir o serviço */
    @Bean
    Queue dlqQueue() {
        return QueueBuilder.durable("video.dlq").build();
    }

    @Bean
    Jackson2JsonMessageConverter messageConverter() {
        return new Jackson2JsonMessageConverter();
    }

    /** Container factory com ACK manual (mesmo padrão do worker) */
    @Bean
    SimpleRabbitListenerContainerFactory manualAckFactory(ConnectionFactory cf) {
        var factory = new SimpleRabbitListenerContainerFactory();
        factory.setConnectionFactory(cf);
        factory.setAcknowledgeMode(AcknowledgeMode.MANUAL);
        factory.setPrefetchCount(1);
        factory.setMessageConverter(messageConverter());
        return factory;
    }
}

File: notification-service/src/main/kotlin/com/fiapql/notification/config/RabbitMqConfig.kt
package com.fiapql.notification.config

import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
import org.springframework.amqp.core.AcknowledgeMode
import org.springframework.amqp.core.Queue
import org.springframework.amqp.core.QueueBuilder
import org.springframework.amqp.rabbit.config.SimpleRabbitListenerContainerFactory
import org.springframework.amqp.rabbit.connection.ConnectionFactory
import org.springframework.amqp.support.converter.Jackson2JsonMessageConverter
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration

@Configuration
class RabbitMqConfig {

    @Bean
    fun dlqQueue(): Queue = QueueBuilder.durable("video.dlq").build()

    // ObjectMapper com módulo Kotlin: necessário para desserializar a data class da mensagem
    @Bean
    fun messageConverter() = Jackson2JsonMessageConverter(jacksonObjectMapper())

    @Bean
    fun manualAckFactory(cf: ConnectionFactory): SimpleRabbitListenerContainerFactory {
        val factory = SimpleRabbitListenerContainerFactory()
        factory.setConnectionFactory(cf)
        factory.setAcknowledgeMode(AcknowledgeMode.MANUAL)
        factory.setPrefetchCount(1)
        factory.setMessageConverter(messageConverter())
        return factory
    }
}
